core/trino-main/src/test/java/io/trino/sql/planner/optimizations/TestAddExchangesPlans.java

```diff
@@ -978,6 +991,241 @@ public void testJoinNotExactlyPartitioned()
+    {
+        MockConnectorFactory connectorFactory = MockConnectorFactory.builder()
+                .withGetColumns(schemaTableName -> ImmutableList.of(
+                        new ColumnMetadata("nationkey", BigintType.BIGINT)))
+                .withGetTableHandle(((session, schemaTableName) -> new MockConnectorTableHandle(
+                        SchemaTableName.schemaTableName("default", "nation"),
+                        TupleDomain.all(),
+                        Optional.of(ImmutableList.of(new MockConnectorColumnHandle("nationkey", BigintType.BIGINT))))))
+                .withName("mock")
+                .build();
+        getQueryRunner().createCatalog("mock", connectorFactory, ImmutableMap.of());
+
+        // Need to use JOIN as parent of UNION ALL to expose replacing remote exchange with local exchange
+        assertDistributedPlan(
+                """
+                         SELECT * FROM (SELECT nationkey FROM nation UNION ALL SELECT nationkey FROM mock.default.nation), region
+                """,
+                noJoinReordering(),
+                output(
+                        join(INNER, join -> join
+                                .left(exchange(REMOTE, REPARTITION, FIXED_ARBITRARY_DISTRIBUTION,
+                                        tableScan("nation"),
+                                        node(TableScanNode.class)))
+                                .right(
+                                        exchange(LOCAL, GATHER, SINGLE_DISTRIBUTION,
+                                                exchange(REMOTE, REPLICATE, FIXED_BROADCAST_DISTRIBUTION,
+                                                        tableScan("region")))))));
+    }
+
+    @Test
+    public void testUnionAllOnInternalCatalog()
+    {
+        // Need to use JOIN as parent of UNION ALL to expose replacing remote exchange with local exchange
+        // TODO: https://starburstdata.atlassian.net/browse/SEP-11273
+        assertDistributedPlan(
+                """
+                         SELECT * FROM (SELECT table_catalog FROM system.information_schema.tables UNION ALL SELECT table_catalog FROM system.information_schema.tables), region
+                """,
+                noJoinReordering(),
+                output(
+                        join(INNER, join -> join
+                                .left(exchange(LOCAL, REPARTITION, FIXED_ARBITRARY_DISTRIBUTION,
+                                        tableScan("tables"),
+                                        tableScan("tables")))
+                                .right(
+                                        exchange(
+                                                LOCAL, GATHER, SINGLE_DISTRIBUTION,
+                                                exchange(REMOTE, REPLICATE, FIXED_BROADCAST_DISTRIBUTION,
+                                                        tableScan("region")))))));
+    }
+
+    @Test
+    public void testUnionAllOnTableScanAndValues()
```

Comment:
consider adding test:
```
                union
                |    |
            proj     ts
            |
         union
        /    \
        ts    ts
```
2 immediate childre, 3 table scans in total. all unions should be translated into `local` exchange